Improve sorting and support grouping for bibliographies

hendrikvanantwerpen opened the pull request:
This pull request introduces improved sorting and support for grouping of bibliographies. I use it myself for a list of publication, which is otherwise very messy. I'm happy to discuss and improve the code, if you feel this is necessary. The `jekyll-scholar-extras` also provide support for grouping, but it introduces an extra dependency, and one that is not easily availble (not in the standard gem repository). I think having this in `jekyll-scholar` itself can be a nice feature.

It includes the following changes:
 * Sort `order` can be mulitlevel, i.e. `order: descending,ascending` will sort the first sort key ascending, and the second sort key descending. If there are more sort keys than order directives, the last item in `order` is used for the remaining keys.
 * Introduce `group_by` and `group_order` settings, to allow grouping of bibliography items.
   * Special care is taken to render proper names for entry `type`s and `month`s. The names for entry types can be customized using the `type_names` setting.
   * Ordering `group_order` is multi-level, just as for sorting.
   * Entry types have no inherit ordering, but one can be specified using `type_order`. Entry types can be merged using `type_aliases`, by default PhD and Master theses are treated as one entry type for grouping.
   * Because the CiteProc renderer sometimes gives different output depending on the previous entry -- e.g. not listing authors of they are the same -- a new instance is created for every group.
   * Limiting entries is disabled when grouping is active.

inukshuk commented:
This is great! Grouping is regularly requested but so far no one has come up with a sufficiently generic solution so I'll be happy to merge this.

Do you have an idea how to make the month names configurable? I'm pretty sure this would be one of the first follow-up feature requests to this.

Also, please document the feature in the readme and add a few test cases. Having to maintain the code down the road I absolutely need the tests to catch regressions. Having said that, this contribution qualifies you for write access to the repository -- if you would be interested to help out, just let me know. Thanks again!

hendrikvanantwerpen commented:
Re documentation and tests: I added two paragraphs to the README and added a feature for grouping.

Re month names: I use the English names from Ruby now, but alternative month names can be specified as a list in the configuration.

Re write access: I'm happy to fix issues with my contribution. However, I don't expect to have much time to be actively involved at the moment. So thanks for the offer, but I don't think I will be useful at the moment.

inukshuk commented:
Awesome, thanks!

Can you fix the Ruby 2 errors (probably use `Hash[]` instead of `to_h` but I didn't look closely)?

hendrikvanantwerpen commented:
Looking into it.

hendrikvanantwerpen commented:
Removed the use of `to_h`. I also moved the ordering of groups to the tag, since hashes are not order-preserving in every Ruby version.

inukshuk commented:
Landed in 5.4.0 / 5.4.1 -- Thanks again!

hendrikvanantwerpen commented:
Cool! You're welcome.
